File: parsers/parse.hs
{-# LANGUAGE GADTs, FlexibleContexts #-}
module Parser where
import Control.Monad
import Lexer

data Transition where
    None :: Transition
    Accept :: Transition
    Shift :: String -> Transition
    Reduce :: String -> Int -> Transition
    deriving (Show, Eq)

data SemanticVal where
    S :: String -> SemanticVal
    I :: Int -> SemanticVal
    Error :: String -> SemanticVal
    --V :: [a] -> SemanticVal
    deriving (Show, Eq)

pos :: Lexer.Token -> (Int, Int)
pos (_, _, p) = p

tokenName :: Lexer.Token -> String
tokenName (Lexer.Name n, _, _) = n
tokenName (Lexer.Skip, _, _) = "(SKIP)"
tokenName (Lexer.Err e, _, _) = "(ERR) " ++ e

parser :: (String -> (String -> Transition)) -> String -> [Lexer.Automata] -> String -> SemanticVal
parser table start automata input =
    case Lexer.lexer automata input of
        Nothing -> Error "Bad lexing"
        Just tokens ->
            case tokenName (last tokens) of
                "\x18" -> actualParser table tokens [start] []
                _ -> actualParser table (tokens ++ [(Lexer.Name "\x18", "\x18", (Lexer.tRow (last tokens), Lexer.tCol (last tokens)))]) [start] []

actualParser :: (String -> (String -> Transition)) -> [Lexer.Token] -> [String] -> [SemanticVal] -> SemanticVal
actualParser _ [] _ _ = Error "Unexpected EOF"
actualParser _ _ [] _ = Error "Empty stack"
actualParser table tokens (top:stack) semstack = do {
    case table top (tokenName (head tokens)) of
        None -> Error ("None " ++ show (pos (head tokens)) ++ ": " ++ top ++ ", " ++ tokenName (head tokens))
        Accept -> if null semstack then S "success" else head semstack
        Shift s -> actualParser table (drop 1 tokens) (s:top:stack) semstack
        Reduce s n -> case table (stack !! max 0 (n - 1)) s of
            Shift str -> actualParser table tokens (str : drop (n - 1) stack) semstack
            _ -> Error ("Bad reduce " ++ show (pos (head tokens)) ++ ": " ++ top ++ ", " ++ tokenName (head tokens))
}
-- actualParser _ tokens _ _ = S (show tokens)

File: parsers/parse.py

STARTNONTERM = "\31"
EOF = "\30"
start = ""
semstack = []
table = {}
actions = {}
errorSets = {}
currentToken = None

class Parser:
    def __init__(self,tks: str):
        self.stack = [start]
        self.lookahead = 0
        self.tkstream = lexer.lex.start(tks)
        global semstack, currentToken
        semstack = []
        currentToken = None

    def parse(self):
        global semstack, errorSets
        while len(self.stack) != 0:
            next = table[self.stack[0]][self.tkstream[self.lookahead]['symbol']]
            #accept
            if next == "accept":
                if len(semstack)>0: return semstack[0]
                else: 
                    print("CRINGE! Parse error: empty sem stack")
                    exit(1)
            #failure
            elif next == None:
                f = None
                state = list(filter(lambda x: x != '', self.stack[0].split(" ")))
                if state[-1] == "●":
                    f = str(errorSets["f" + state[0]])
                else:
                    f = str(errorSets[state[state.index("●") + 1]])
                print("Parse error " + str(self.tkstream[self.lookahead]['pos']) \
                    + ": Found " + str(self.tkstream[self.lookahead]['symbol']).replace('\x18', '\\x18') \
                    + ":" + str(self.tkstream[self.lookahead]['lexeme']) + ", Expected " + f)
                exit(1)
            #shift/goto
            elif isinstance(next,str):
                self.stack = [next] + self.stack
                self.lookahead += 1
                global currentToken
                currentToken = self.tkstream[self.lookahead - 1]["lexeme"]
            #reduce
            else:
                if self.stack[0] in actions.keys():
                    actions[self.stack[0]]()
                self.stack = [table[self.stack[next[1]]][next[0]]] + self.stack[next[1]:]
        print("Parse Error: unexpected EOF")
        exit(1)
